### PR TITLE
Support for FNIRSI FNB58 is here!

### LINKED ISSUE #2: How to run it with FNB58 ?
Fnirsi released the new version of their usb meters, Fnb58. How can I use your code with it ? What modifications do I need to make? At the moment, "device not found" is thrown. Have you tested it with the new device ? Thanks

### PR DESCRIPTION
**Pull request by didim99**
Fixes #2, and also probably fixes FNB48 freezes when program exiting (tested couple dozens of times but needs much more testing to be sure).

### CODE REVIEW

**Comment by baryluk** [comment]
Thank you. I tested this on C1 and FNB48, and on C1 it works perfectly, including exiting does not freeze the screen.

But on FNB48 it has some weirdness, on exit it does freeze the FNB48 screen, and running script again without resetting the FNB48, the first sample is weird:

FNB48:

```
user@debian:~/fnirsi-utils$ sudo ./fnirsi_logger.py
....
1667765030.189 1 0.00000 0.00000 1.404 1.191 23.699 0.000000 0.000038
1667765030.199 2 0.00000 0.00000 1.403 1.181 23.709 0.000000 0.000038
1667765030.209 3 0.00000 0.00000 1.411 1.197 23.708 0.000000 0.000038
1667765030.219 0 0.00000 0.00000 1.410 1.194 23.697 0.000000 0.000038
1667765030.229 1 0.00000 0.00000 1.404 1.185 23.698 0.000000 0.000038
1667765030.239 2 0.00000 0.00000 1.403 1.183 23.698 0.000000 0.000038
1667765030.249 3 0.00000 0.00000 1.411 1.195 23.698 0.000000 0.000038
^C....
KeyboardInterrupt

(run again while the FNB48 is still attached)

user@debian:~/fnirsi-utils$ sudo ./fnirsi_logger.py 

timestamp sample_in_packet voltage_V current_A dp_V dn_V temp_C_ema energy_Ws capacity_As
1667765050.059 0 188.74370 170.39408 51.840 0.000  0.000 321.608091 1.703941
1667765050.069 1 0.00000 0.00000 0.000 0.000  0.000 321.608091 1.703941
1667765050.079 2 0.00000 0.00000 0.000 0.000  0.000 321.608091 1.703941
1667765050.089 3 0.00000 0.00000 0.000 0.000  0.000 321.608091 1.703941
1667765050.065 0 0.00000 0.00000 1.376 1.129  2.420 321.608091 1.703941
1667765050.075 1 0.00000 0.00000 1.373 1.125  4.598 321.608091 1.703941
1667765050.085 2 0.00000 0.00000 1.376 1.130  6.558 321.608091 1.703941
1667765050.095 3 0.00000 0.00000 1.376 1.129  8.322 321.608091 1.703941
```

This is weird, as I remember I had a code to ignore first data sample, as it was some total random values (probably not random, but they were probably encoding something else than normal samples - you can see above that dp_V dn_V temp_C are also wrong in the first data packet in samples 1,2,3).

There is probably nothing wrong with your code, but I will also do a bit more testing, including original code.

File: fnirsi_logger.py [hunk 1]
```diff
@@ -96,12 +98,15 @@ def main():
     assert ep_out
 
     ep_out.write(b"\xaa\x81" + b"\x00" * 61 + b"\x8e")
-    ep_out.read(size_or_buffer=64)
+    if is_fnb58:
```
Comment (baryluk) [inline fnirsi_logger.py:101]: I think this should be `if is_fnb58 or is_fnb48:`

File: fnirsi_logger.py [hunk 2]
```diff
@@ -24,11 +24,13 @@
 
 def main():
     # find our device
+    is_fnb58 = False
     dev = usb.core.find(idVendor=VID, idProduct=PID_FNB48)
     if dev is None:
         dev = usb.core.find(idVendor=VID, idProduct=PID_C1)
     if dev is None:
         dev = usb.core.find(idVendor=VID_FNB58, idProduct=PID_FNB58)
+        is_fnb58 = True
```
Comment (baryluk) [inline fnirsi_logger.py:33]: Lets make this:

```python3
if dev is not None:
  is_fnb58 = True
```

And introduce similar one for fnb48.

For now these bools are fine, but using enum (i.e. `class DeviceModel(enum.Enum)` , or `DeviceQuirks`) is also an option.

File: fnirsi_logger.py (hunk 1 above)
Comment (didim99) [reply fnirsi_logger.py]: Probably not, I looked more closely on HID packet and determined what are first byte is, probably it is a payload type. All data packets starts from 0x04 byte, some other packets (sometimes first, sometimes not) starts from 0x03 byte and these packets not is data samples (status report?). We just can ignore all non-data packets and don't use workarounds (crutch?) such as `first_data_point`. See last commit.

File: fnirsi_logger.py (hunk 2 above)
Comment (didim99) [reply fnirsi_logger.py:33]: I suggest to use some abstraction such as my [`DeviceInfo`](https://github.com/didim99/witrn-ui/blob/84cfc265df96261ff1ff0b6b52883d1398291538/driver/protocol.py#L7) if we need it. In future I plan to merge all open source solutions related to USB-meters and other similar equipment (such as loads and programmable PSUs) in one universal data logging/control tool and these abstractions may be useful for common extensible driver architecture.

File: fnirsi_logger.py (hunk 1 above)
Comment (baryluk) [reply fnirsi_logger.py]: Yes, using just `0x04` is good approach.

I do not know why I didn't notice this before in my RE checks.

I am guessing that `0x03` is some device info like model name, serial number, and firmware version, etc. But it is just a guess at the moment.

File: fnirsi_logger.py (hunk 2 above)
Comment (baryluk) [reply fnirsi_logger.py:33]: Yes. Now with `data[0] == 0x04` we can skip this.

Resolving.

**Pull request merged**